File: scala/src/main/scala/sodium/BehaviorSink.scala
package sodium

/**
  * A cell that allows values to be pushed into it, acting as an interface between the
  * world of I/O and the world of FRP. Code that exports CellSinks for read-only use
  * should downcast to [[Behavior]].
  *
  * @constructor Construct a writable cell with the specified initial value. If multiple values are
  * sent in the same transaction, the specified function is used to combine them.
  */
final class BehaviorSink[A](initValue: A, f: (A, A) => A) extends Behavior[A](new StreamSink[A](f), Some(initValue)) {

  /**
    * Construct a writable cell with the specified initial value.
    * If multiple values are sent in the same transaction, the last one is used.
    */
  //TODO remove this trick, what we really want is the constructor StreamSink.this() not a copy of it.
  def this(initValue: A) =
    this(
      initValue,
      (_: A, _: A) =>
        throw new RuntimeException(
          """send() called more than once per transaction, which isn't allowed. Did you want to combine the events?
        |Then pass a combining function to your StreamSink constructor.""".stripMargin)
    )

  /**
    * Send a value, modifying the value of the cell. send(A) may not be used inside
    * handlers registered with [[Stream.listen(handler:A=>Unit):sodium\.Listener* Stream.listen(A=>Unit)]] or
    * [[Behavior.listen Behavior.listen(A=>Unit)]].
    * An exception will be thrown, because BehaviorSink is for interfacing I/O to FRP only.
    * You are not meant to use this to define your own primitives.
    *
    * @param a Value to push into the cell.
    */
  def send(a: A): Unit = {
    str match {
      case s: StreamSink[A] => s.send(a)
      case _                =>
    }
  }
}

File: dotnet/src/Sodium/Sodium.Core.Frp/BehaviorSink.cs
using System;

namespace Sodium.Frp
{
    /// <summary>
    ///     A behavior that allows values to be pushed into it, acting as an interface between the world of I/O and the world of
    ///     FRP.  Code that exports instances of <see cref="BehaviorSink{T}" /> for read-only use should downcast to
    ///     <see cref="Behavior{T}" />.
    /// </summary>
    /// <typeparam name="T">The type of values in the behavior sink.</typeparam>
    public class BehaviorSink<T> : Behavior<T>
    {
        private readonly StreamSink<T> streamSink;

        internal BehaviorSink(T initialValue)
            : this(new StreamSink<T>((left, right) => right), initialValue)
        {
        }

        internal BehaviorSink(T initialValue, Func<T, T, T> coalesce)
            : this(new StreamSink<T>(coalesce), initialValue)
        {
        }

        private BehaviorSink(StreamSink<T> streamSink, T initialValue)
            : base(streamSink, initialValue) => this.streamSink = streamSink;

        internal void SendImpl(T a) => this.streamSink.SendImpl(a);
    }
}